### ISSUE TITLE
[question] Does apex use white/black lists?

### ISSUE THREAD

**Issue by vadimkantorov** [NVIDIA/apex#656]
If I understand well, now apex does not special case SoftMax. Is it true?

If I understand well, only BatchNorm and RNN params are currently special cased.

**Comment by mcarilli** [comment]
What do you mean special cased?  The full list of functions that Amp patches can be found here:
https://github.com/NVIDIA/apex/tree/master/apex/amp/lists
Amp doesn't patch batchnorm at all, actually.  With cudnn, batchnorm can safely handle either fp32 or fp16 inputs, as long as the weights remain fp32, which they do with opt_level="O1".

**Comment by vadimkantorov** [comment]
I couldn't find amp.init() call (that does the patching) from inside amp.initialize(...)...

**Comment by vadimkantorov** [comment]
Ah, got it. It's clear now.

**Comment by vadimkantorov** [comment]
Here's that call: https://github.com/NVIDIA/apex/blob/master/apex/amp/_initialize.py#L235
